### ISSUE TITLE
Decoding the Trigger block for TestBeam data

### ISSUE THREAD

**Issue by pfs** [CMS-HGCAL/cmssw#77]
After a discussion of the current needs and also the plans for future system validation it seems adequate to start by implementing a simple unpacking and monitoring of the DQM blocks. Thus trigger data will be stored for the moment as raw data in the FEDRawDataCollection. 
Once the this is in place and validated we'll think about possible filling a CMSSW data format which can be used for persistent storage and filling of the NANOAOD. The plan outlined is the following:

- [x] Toni/Indra : select a good run to start with
- [x] Pedro/Huilin: will implement the reading of the raw and FW-unpacked raw blocks and it's storage in the FEDRawDataCollection
- [x] Toni/Indra: will provide the standalone unpackers, fill the DQM plots needed to count errors etc.
- [x] Pedro/Huilin: will create a DQMClient to read directly the relevant entry in the FEDRawDataCollection where a call to  "standalone" unpackers can be made

**Comment by indra-ehep** [comment]
I will recommend run number 1695587722. It has about 0.8 billion physics triggered events which is very useful for DQM plots. This is probably the longest run collected during Septemeber. I found no serious corruption in trigger data and all trigger cells had been active.  There will be very few cases (~55) where there will be header mismatches comparing from two sources.  However, there can be a significant number of mismatches while comparing the energy (or location) of rawdata and unpacker output. All these are good to have meaningful DQM plots.

If we do not want to accept any errors for ECON-T raw and unpacker output (energy and location), then we can use run number 1695855233. Please note most of the DQM plots could be empty, since it has ~58K events with calibration trigger.

**Comment by tsculac** [comment]
> I will recommend run number 1695587722. It has about 0.8 billion physics triggered events which is very useful for DQM plots. This is probably the longest run collected during Septemeber. I found no serious corruption in trigger data and all trigger cells had been active. There will be very few cases (~55) where there will be header mismatches comparing from two sources. However, there can be a significant number of mismatches while comparing the energy (or location) of rawdata and unpacker output. All these are good to have meaningful DQM plots.
> 
> If we do not want to accept any errors for ECON-T raw and unpacker output (energy and location), then we can use run number 1695855233. Please note most of the DQM plots could be empty, since it has ~58K events with calibration trigger.

I agree with Indra. Let us know once you have the raw blocks ready so that we can work on implementing our unpackers.

**Comment by pfs** [comment]
@indra-ehep @tsculac mind sharing the slide describing the raw data format in this thread. 
Also below is a printout of a random event for what i would think is the raw and FW unpacked raw trigger data. Mind checking if they make sense?

```
Raw data
	 0xeaff2b40fbffff0
	 0x49522204409a8204
	 0x810242708102040
	 0x11a00000810c0000
	 0x1e9ff3a41fbfcff0
	 0x4951a204409a8204
	 0x810236708106040
	 0x51200000810f0000
	 0x2e6ff3e42fbeaff0
	 0x69522204409a8204
	 0x81023a714285040
	 0x91a1000081100000
	 0x3e9ff3943fbfbff0
	 0x6952a204409a8204
	 0x81023a608282040
	 0xd1200000810f0000
	 0x4eeff2f44fbffff0
	 0x4992a204409a8204
	 0x810236608102040
	 0xd0200000810c0000
	 0x5eaff2b45fbffff0
	 0x6952a304409a8204
	 0x810236608102040
	 0xd1200000810a0000
	 0x6edff2b46fbf8ff0
	 0x6952a304409a8204
	 0x81023c708485040
	 0x91210000810a0000
	 0x7e8a8e967e9e9694
	 0x4be7b13221a11c46
	 0x763235689b432505
	 0x2a2200009fa40000
	 0xecfbe140fb3d691
	 0x6919a604009c020e
	 0x8e022e33538d261
	 0x19140000948b0000
	 0x1edff2e41fbfdfb0
	 0x4914a404409b020a
	 0x81023c709017040
	 0x14230000888e0000
	 0x2edff2642fbffff0
	 0x4952a304409a8204
	 0x810242808102040
	 0x11a3000081100000
	 0x3e9ff2b43fbffff0
	 0x6953a404409a8204
	 0x81063a708102040
	 0x12220000810c0000
	 0x4e9ff2a44fbf8ff0
	 0x49522204409a8204
	 0x810238608485040
	 0x51210000810e0000
	 0x5e6ff2a45fbffff0
	 0x4912a204409a8204
	 0x810238608102040
	 0xd1a10000810a0000
	 0x6e5ff2a46fbffff0
	 0x69122104409a8204
	 0x810238708102040
	 0x11200000810a0000

FW-unpacked raw data
	 0x81b041a281b04182
	 0x81f0432581f04382
	 0x8204252482304535
	 0x8263872282704782
	 0x82a4698282b04982
	 0x82f40b8282f18b82
	 0x85b041a285b04182
	 0x85f0432585c0c382
	 0x8603652386304535
	 0x8673a6a286704782
	 0x86a4498286b04982
	 0x86e40b8286f1eb82
	 0x89b041a389a0a182
	 0x89f0432589e0a382
	 0x8a03a4a48a304535
	 0x8a73c7228a704782
	 0x8ab469828ab04982
	 0x8ae42b828af20b05
	 0x8db041a38da0a182
	 0x8df043258df04382
	 0x8e03a5258e304535
	 0x8e7366a28e704782
	 0x8ea449828eb04982
	 0x8ed40b828ef1eb82
	 0x91b041a291b04182
	 0x91f0432691f04382
	 0x920365a592304535
	 0x9263672292704782
	 0x92b4098292b04982
	 0x92f40b8292f18b82
	 0x95b041a395b04182
	 0x95f0432595f04382
	 0x9603652596304535
	 0x9663672396704782
	 0x96a4498296b04982
	 0x96f40b8296f14b82
	 0x99b041a399a12182
	 0x99f0432599c0a382
	 0x9a03c5a59a304535
	 0x9a63c6a39a704782
	 0x9aa449829ab04982
	 0x9af42b829af14b82
	 0x9da8c1b29dad01a1
	 0x9dc4632f9dd64306
	 0x9e35654f9e150542
	 0x9e6406319e62c69c
	 0x9eaa89199ea7e9a3
	 0x9ed44b1d9ed48b26
	 0x81a381a381b4e188
	 0x81f0432481d1a382
	 0x8232e5b382126538
	 0x826186268260c782
	 0x8286498282a52807
	 0x82d28b8282d16b8d
	 0x85b041a285b40182
	 0x85f0432485d2e382
	 0x8603c5a986304536
	 0x866386a486704782
	 0x86b5098286a22985
	 0x86e46b8286f1cb82
	 0x89b041a289b04182
	 0x89f0432589f04382
	 0x8a0425a58a304535
	 0x8a6406a38a704782
	 0x8a9469828ab04982
	 0x8ae46b828af20b82
	 0x8db041a38db04182
	 0x8df0c3258df04382
	 0x8e03a5278e304535
	 0x8e6386a48e704782
	 0x8ea489828eb04982
	 0x8ef44b828ef18b82
	 0x91b041a291a12182
	 0x91f0432591c0a382
	 0x9203852492304535
	 0x926326a292704782
	 0x92a4498292b04982
	 0x92e42b8292f1cb82
	 0x95b041a295b04182
	 0x95f0432495f04382
	 0x960384a596304535
	 0x9663672296704782
	 0x96a4698296b04982
	 0x96e42b8296f14b82
	 0x99b041a399b04182
	 0x99f0432499f04382
	 0x9a0384a49a304535
	 0x9a6386a19a704782
	 0x9aa449829ab04982
	 0x9ae40b829af14b82
```

**Comment by tsculac** [comment]
Hi @pfs . The slide 6 in this presentation [1] is showing raw data format of ECON-T. We would need to have the header as well to determine the size of the package. For example on the attached slide the header of ECON-T raw data is 0xfecafecafecafe24 which tells us that there are 2*16 + 4 = 36 words in the package.

[1][https://indico.cern.ch/event/1327255/contributions/5585405/attachments/2718230/4721885/HGCAL_TPG_TestBeam_230920.pdf](https://indico.cern.ch/event/1327255/contributions/5585405/attachments/2718230/4721885/HGCAL_TPG_TestBeam_230920.pdf)

**Comment by pfs** [comment]
Ok thanks a lot. for the size in CMSSW the FEDRawDataCollection is a vector of char per FED so you get the size from the size of the vector itself. Unless you see any drawback we could getaway without storing the header word.

**Comment by tsculac** [comment]
The only potential use I can see is for debugging in the beginning. Otherwise I agree there is no need to save the header.

**Comment by indra-ehep** [comment]
> Ok thanks a lot. for the size in CMSSW the FEDRawDataCollection is a vector of char per FED so you get the size from the size of the vector itself. Unless you see any drawback we could getaway without storing the header word.

@pfs @tsculac 
If it is not too much of storage issue, I would vote to keep the ECON-T header for the moment. It is can be dropped later whenever (??) we are confident that there is no error in the data :wink: : . There are certain information that are only present in header/trailer and not in paylaod. E.g.

1. EventId
2. Sequence counter
3. bxId
4. Orbit counter, event counter, bxcounter
5. L1aType
6. validity of payload 0x55 and 0xaa

Please note, the above list is not the complete one.
In September beamtest data there are several issues that can only be traced by checking and validating the headers.

**Comment by indra-ehep** [comment]
[Beamtest_readout_format.pdf](https://github.com/CMS-HGCAL/cmssw/files/12858882/Beamtest_readout_format.pdf)

**Comment by indra-ehep** [comment]
> [Beamtest_readout_format.pdf](https://github.com/CMS-HGCAL/cmssw/files/12858882/Beamtest_readout_format.pdf)

I am using this file to find out the issues in the ECON-T data.

**Comment by hqucms** [comment]
Hi @indra-ehep @tsculac -- I created the branch [dev/hgcal-condformat-HGCalNANO-13_2_0_pre3/econt_unpacker](https://github.com/CMS-HGCAL/cmssw/tree/dev/hgcal-condformat-HGCalNANO-13_2_0_pre3/econt_unpacker) with the reading of the raw trigger link files and the creation of `FedRawDataCollection`. In that branch there is also a placeholder DQM client [HGCalTriggerClient.cc](https://github.com/CMS-HGCAL/cmssw/blob/dev/hgcal-condformat-HGCalNANO-13_2_0_pre3/econt_unpacker/DQM/HGCal/plugins/HGCalTriggerClient.cc#L90-L109) where the unpacking of the `FedRawDataCollection` can be added. 

The test cmsRun command is:

```bash
cmsRun ${CMSSW_BASE}/src/EventFilter/HGCalRawToDigi/test/tb_raw2reco.py mode=slinkfromraw slinkBOE=0x55 cbHeaderMarker=0x7f econdHeaderMarker=0x154 applyFWworkaround=False inputFiles=/eos/cms/store/group/dpg_hgcal/tb_hgcal/2023/BeamTestSep/HgcalBeamtestSep2023/Relay1695587722/Run1695587722_Link1_File0000000000.bin,/eos/cms/store/group/dpg_hgcal/tb_hgcal/2023/BeamTestSep/HgcalBeamtestSep2023/Relay1695587722/Run1695587722_Link2_File0000000000.bin fedId=1,2 inputTrigFiles=/eos/cms/store/group/dpg_hgcal/tb_hgcal/2023/BeamTestSep/HgcalBeamtestSep2023/Relay1695587722/Run1695587722_Link0_File0000000000.bin output=Run1695587722_Link2_File0000000000 conditions=default dqmOnly=True runNumber=1695587722 maxEvents=10 
```

Let me know if you have any questions.

**Comment by indra-ehep** [comment]
Dear @hqucms,

Thanks for creating the branch and template files. I have a query.

First I have installed CMSSW-13_2_0_pre3 project using cmsrel, then applied 
git cms-merge-topic -u CMS-HGCAL:dev/hgcal-condformat-HGCalNANO-13_2_0_pre3/econt_unpacker
to include your development followed by compile command.
Then  cmsRun command is applied.
Please find the command printouts attached.

Query : 
1. What is the expected output file ? Should I expect a NanoAOD.root files to be produced with the ECON-D and dummy ECON-T tree branches ?
In lxplus, I have notices two DQM files DQM_V0001_HGCAL_R1695587722.root.origin and DQM_V0001_HGCAL_R1695587722.root and two txt files level0_deadch.txt and level0_calib_params.txt.

Or do I have to run some other cmsRun command to get the nanoAOD tree for the concentrators ?

**Comment by indra-ehep** [comment]
[lxplus_check.log](https://github.com/CMS-HGCAL/cmssw/files/13217644/lxplus_check.log)

**Comment by indra-ehep** [comment]
Dear @hqucms 

I now see that the motivation of this task is first to look at the DQM plots. So for this task I do not explicitly request for nanoAOD branch. But can you tell me how to produce such objects for one run ? For instance, as it is done for the results at /eos/cms/store/group/dpg_hgcal/tb_hgcal/2023/CMSSW/ReReco_Oct10/.

best 

Indra

**Comment by pfs** [comment]
Hi @indra-ehep 
if you take one of the relays you'll find a subdirectory called logs with files called `cmds*.log`
If you look to the content of those files you'll find the commands which can be used locally to run the RECO, DQM and NANOAOD steps. I believe you want to first run RECO for unpacking and then DQM, or maybe it may be best to concatenate the two in a single file to do directly RECO-DQM.

**Comment by pfs** [comment]
ah looking to the log you provided now, so you can already run this in a single step and receive the block in the DQM trigger client! :)

**Comment by hqucms** [comment]
Hi @indra-ehep -- sorry for my late reply, I am travelling these days.

The command for each step (RECO, DQM, NANO) can be found as in the log files as Pedro mentioned. 

If you only need to produce the DQM file (e.g., for quick testing), here is an example command:
https://github.com/CMS-HGCAL/cmssw/issues/77#issuecomment-1765376192

And the `DQM_V0001_HGCAL_R1695587722.root` output is the DQM file with histograms/hexplots inside, which you can inspect with plain ROOT.

For producing NanoAODs -- I think for that we need first to define a data format to store the all the relevant information we want to extract from the ECON-T blocks, and then write a EDProducer to use the ECON-T unpacker that you develop to produce these data objects, and the update the NanoAOD part to write out these objects.

**Comment by indra-ehep** [comment]
Dear @pfs and @hqucms,

@pfs : Thank you for pointing the file to find the command for producing NanoAOD.
@hqucms : No worries have a safe travel. Sooner or later we will need to fill those NanoAOD for ECON-T. As a first thought it should contain for each trigger cells.

a. Energy
b. bxindex 
c. local Electronic location
d. global geometric location
e. Calibration
f. wafer  number
g. layer number 

However please wait for few days, the memebrs for ECON-T NanoAOD will be more clear once the code of current task is implemented.

**Comment by indra-ehep** [comment]
@hqucms , @pfs  Is it possible to access members of HGCalSlinkFromRaw in  DQM client [HGCalTriggerClient.cc](https://github.com/CMS-HGCAL/cmssw/blob/dev/hgcal-condformat-HGCalNANO-13_2_0_pre3/econt_unpacker/DQM/HGCal/plugins/HGCalTriggerClient.cc#L90-L109) ?
It is requested for some eventId validations through class SlinkBoe.

**Comment by indra-ehep** [comment]
@hqucms , @pfs  Run 1695496516 has total 335082 events, which are divided in three bin files Run1695496516_Link0_File0000000000.bin, Run1695496516_Link0_File0000000001.bin and Run1695496516_Link0_File0000000002.bin. When I run separately on each of then sum using hadd I get about 10% in the histogram count for the attached code. How to record for each event instead of skipping 90% in the histogram ?
 
[HGCalTriggerClient.cc.csv](https://github.com/CMS-HGCAL/cmssw/files/13307586/HGCalTriggerClient.cc.csv)

**Comment by hqucms** [comment]
Hi @indra-ehep -- can you try changing the prescale from 1000 to 1 for the DQM in these lines:
https://github.com/hqucms/cmssw/blob/1fe99b7d6b2cd26caf3c6abb5eba33f11937438f/EventFilter/HGCalRawToDigi/test/tb_raw2reco.py#L300-L301

**Comment by indra-ehep** [comment]
Thank @hqucms it worked.

**Comment by indra-ehep** [comment]
@pfs and @hqucms : Please check https://github.com/CMS-HGCAL/cmssw/pull/86 and merge, if satisfied. Please note that it can reproduce the results of standalone trigger block decoding.

**Comment by adavidzh** [comment]
Thanks @indra-ehep. After merging #86 will the decoding of both STC and BC be available offline?

**Comment by indra-ehep** [comment]
@adavidzh : Not yet. We are yet to fill the nanoaod branches, which is the next step. In addition we are yet to implement the BC. But we are closer than before.

**Comment by pfs** [comment]
Let's discuss this live today - i have added Indra's contribution to https://indico.cern.ch/event/1346632/#4-progress-in-decoding-the-tri

**Comment by adavidzh** [comment]
> Let's discuss this live today 

I'm afraid that I am taken by the ESR Part 1 Session 2.

**Comment by indra-ehep** [comment]
> Hi @indra-ehep @tsculac -- I created the branch [dev/hgcal-condformat-HGCalNANO-13_2_0_pre3/econt_unpacker](https://github.com/CMS-HGCAL/cmssw/tree/dev/hgcal-condformat-HGCalNANO-13_2_0_pre3/econt_unpacker) with the reading of the raw trigger link files and the creation of `FedRawDataCollection`. In that branch there is also a placeholder DQM client [HGCalTriggerClient.cc](https://github.com/CMS-HGCAL/cmssw/blob/dev/hgcal-condformat-HGCalNANO-13_2_0_pre3/econt_unpacker/DQM/HGCal/plugins/HGCalTriggerClient.cc#L90-L109) where the unpacking of the `FedRawDataCollection` can be added.
> 
> The test cmsRun command is:
> 
> ```shell
> cmsRun ${CMSSW_BASE}/src/EventFilter/HGCalRawToDigi/test/tb_raw2reco.py mode=slinkfromraw slinkBOE=0x55 cbHeaderMarker=0x7f econdHeaderMarker=0x154 applyFWworkaround=False inputFiles=/eos/cms/store/group/dpg_hgcal/tb_hgcal/2023/BeamTestSep/HgcalBeamtestSep2023/Relay1695587722/Run1695587722_Link1_File0000000000.bin,/eos/cms/store/group/dpg_hgcal/tb_hgcal/2023/BeamTestSep/HgcalBeamtestSep2023/Relay1695587722/Run1695587722_Link2_File0000000000.bin fedId=1,2 inputTrigFiles=/eos/cms/store/group/dpg_hgcal/tb_hgcal/2023/BeamTestSep/HgcalBeamtestSep2023/Relay1695587722/Run1695587722_Link0_File0000000000.bin output=Run1695587722_Link2_File0000000000 conditions=default dqmOnly=True runNumber=1695587722 maxEvents=10 
> ```
> 
> Let me know if you have any questions.

@hqucms @pfs Can you please suggest the modified arguments for "cmsRun" command when the data collection has been performed via DTH ? For instance corresponding to run : 1695573756 as discussed during today's DPG rawdata meeting.

**Comment by hqucms** [comment]
Hi @indra-ehep -- you can find the commands in the log files on eos, e.g., this one:
 - `/eos/cms/store/group/dpg_hgcal/tb_hgcal/2023/CMSSW/ReReco_Oct10/Run1695573756/30d839b6-6d27-11ee-8957-fa163e8039dc/logs/cmds_*.log`
 
 The main changes needed for the DTH runs are:
 ```
 inputFiles= ...Link1... .bin
 fedId=1
 conditions=DTH
 ```

**Comment by indra-ehep** [comment]
@pfs I have applied the per module error count for ECON-T following the discussion of the DPG rawdata meeting. 
The attached results corresponds to the right side plot of slide 11 of https://indico.cern.ch/event/1346632/ .
The updated code has been pushed again.

![Event_errors_2023-11-15_22-12-45](https://github.com/CMS-HGCAL/cmssw/assets/77208636/ec3d99a9-8e58-4e78-bbbf-6308b99edcff)
![PerModuleError_2023-11-15_22-11-36](https://github.com/CMS-HGCAL/cmssw/assets/77208636/720709e4-5b89-42ae-b183-636afdb61a31)

**Comment by indra-ehep** [comment]
@pfs , @hqucms Do you know who will merge the changes  of https://github.com/CMS-HGCAL/cmssw/pull/86 ?

**Comment by pfs** [comment]
Hi @indra-ehep I merged. Let's go with this baseline, thanks a lot. I ticked all the boxes as the DQMClient makes use of a "standalone unpacker" to fill the plots and such. 

@indra-ehep @hqucms @tsculac @adavidzh   next step is to use the "standalone unpacker" to fill some SoA collection so that it can be used by alpaka and also by nanoaod. We didn't discuss these details in yesterday's raw data meeting. Shall we keep the same issue and update the checklist of steps to fill?
